### PR TITLE
Fix handling of the root path in WasiFS resulting in a bad inode structure

### PR DESCRIPTION
**Pull request by syrusakbary**
* Fix renaming a directory that was created with mkdir
* Fix inodes' name becoming stale after a rename
* Add test for mkdir -> rename -> mkdir
* Fix opening files in the presence of a non-default CWD
* Fix deadlock when base FD is not / or .

# Description

### CODE REVIEW

File: lib/wasix/src/fs/mod.rs [hunk 1]
```diff
@@ -1255,13 +1262,31 @@ impl WasiFs {
...
+            let kind = base_inode.kind.read().unwrap();
+
+            // HACK: We preopen '/' with an alias of '.' by default in `prepare_webc_env`. If wasix-libc
+            // picks that up as the base FD, we want to do the same thing we do when the base is '/'
+            if *name == "."
+                && matches!(kind.deref(), Kind::Dir { path, .. } if path.as_os_str().as_encoded_bytes() == b"/")
+            {
+                drop(name);
+                drop(kind);
+                start_inode = base_inode;
```
[the remark this answers is not in the document]
Comment (theduke) [reply lib/wasix/src/fs/mod.rs:1281]: Of course the `.` hack shouldn't exist at all, but alas, I'm not sure if we can remove it.

[the remark this answers is not in the document]
Comment (Arshia001) [reply lib/wasix/src/fs/mod.rs:1281]: No, we almost never need to use the CWD. All syscalls (excpet proc_exec2 iirc) receive a base fd and a path relative to that fd, so the CWD doesn't factor into path resolution.

**Review by theduke — APPROVED** [review]

**Comment by syrusakbary** [comment]
I'll force-merge since macos tests seem to be on a halt atm

**Comment by Arshia001** [comment]
They started running once I restarted them, albeit very slowly and one-by-one. I'll keep an eye on the CI pipeline for possible failures we may want to correct in a later PR.

**Pull request merged**